### PR TITLE
Use null coalescing in many more places

### PR DESCRIPTION
**Pull request by stephentoub**
I previously enabled all of the IDExx rules around null coalescing, but that still left a large number of manual checks in the codebase, so here I used regexes to find and fix up many more occurences.  @mavasani, is there an analyzer/fixer for this, or if not, have we considered one? For the most part, this is style, cleaning up around 5K lines, but a non-trivial number do result in more efficient accesses.

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

**Comment by mavasani** [comment]
@stephentoub We don't support such a analyzer/code fix/refactoring, but we do have a tracking issue for this feature request: https://github.com/dotnet/roslyn/issues/32985.

@CyrusNajmabadi @mikadumont do you know if we have received requests for this feature in the past?

**Comment by CyrusNajmabadi** [comment]
The only reason we didn't do this was because we weren't certain the semantics were identical, and we were worried about power. E.g. the code before only assigns in the case of null. The code after always assigns. 

If ST is ok with this though, then that's a strong indication this is fine and we should just do this :-)

**Comment by CyrusNajmabadi** [comment]
That's fascinating. I missed that in the spec:

> Which follows the existing semantic rules for compound assignment operators (§11.18.3), except that we elide the assignment if the left-hand side is non-null. 

Thanks for making me aware of that. We'll def add this now!

**Comment by CyrusNajmabadi** [comment]
Updated analyzer/fixer here: https://github.com/dotnet/roslyn/pull/62191

File: src/mono/System.Private.CoreLib/src/System/TypeSpec.cs [hunk 1]
```diff
@@ -312,18 +304,12 @@ private void AddName(string type_name)
...
             }
         }
 
-        private void AddModifier(IModifierSpec md)
-        {
-            if (modifier_spec == null)
-                modifier_spec = new List<IModifierSpec>();
-            modifier_spec.Add(md);
-        }
+        private void AddModifier(IModifierSpec md) => modifier_spec ??= new List<IModifierSpec>();
```
Comment (marek-safar) [inline src/mono/System.Private.CoreLib/src/System/TypeSpec.cs:312]: @stephentoub was it intentional that you also removed the Add call?

Comment (stephentoub) [reply src/mono/System.Private.CoreLib/src/System/TypeSpec.cs:312]: Definitely not.  Sorry.   I'll fix.

Comment (stephentoub) [reply src/mono/System.Private.CoreLib/src/System/TypeSpec.cs:312]: https://github.com/dotnet/runtime/pull/72605

**Pull request merged**